### PR TITLE
Export the plugin compatibility flag in Update Site REST API

### PR DESCRIPTION
**Pull request by LinuxSuRen**
We wish I can upgrade all compatible plugins by [jcli](https://github.com/jenkins-zh/jenkins-cli).
See also the [PR link](https://github.com/jenkins-zh/jenkins-cli/pull/258).

### Proposed changelog entries

* Export the plugin compatibility flag in Update Site REST API
* ...

### Submitter checklist

- [ ] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@romenrg

### CODE REVIEW

**Comment by LinuxSuRen** [comment]
@jenkinsci/code-reviewers I was wondering if anyone could take some time to help to review my PR. Thanks!

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
The concept is fine per se, but we should think about better field aliases in REST APIs. Not all methods are comprehensible here IMHO

**Comment by LinuxSuRen** [comment]
@oleg-nenashev , thanks for your response. So, your suggestion is creating a new annotation for Restful API. Right?

**Comment by oleg-nenashev** [comment]
Nope, @Exported already supports named aliases IIRC

**Comment by LinuxSuRen** [comment]
I hope that I can get the reason why a plugin is not compatible. All methods here are the potential reason. Does this make sense?

**Review by yJunS — COMMENTED** [review]
I think if the API can have 2 methods will be very good.

- check all compatible plugins and show them all.
- check if it is a compatible plugin from transfer a plugin name

**Comment by LinuxSuRen** [comment]
Any updates?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Sorry, my comment about naming has not been addressed yet. Here is the list of available attributes: https://stapler.kohsuke.org/apidocs/org/kohsuke/stapler/export/Exported.html . Also `Exported` should be removed from the methods with `cache` arguments, parameterized getters are not exported IIRC

**Comment by LinuxSuRen** [comment]
@oleg-nenashev Thanks for giving the details. I removed the wrong usage of `Export` annotation. In my point, I think it's understandable.
```
      "compatible" : true,
      "compatibleSinceVersion" : null,
      "compatibleWithInstalledVersion" : true,
```
I was wordering if you or other people could suggest a better name.

**Comment by oleg-nenashev** [comment]
> I removed the wrong usage of Export annotation. In my point, I think it's understandable.

But what about `isForNewerHudson`, `isForNewerJava`, `isNeededDependenciesForNewerJava`? I think it would need some clarification. Also, I doubt we want to expose "Hudson" in newly introduced REST API (can be solved by introducing a new `isForNewerHudson()` method and deprecating the existing one)

**Comment by fcojfernandez** [comment]
I 100% agree with Oleg.

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 1]
```diff
@@ -1125,7 +1126,8 @@ else if (!current.isEnabled()) {
 
             return deps;
         }
-        
+
+        @Exported
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: Should probably be exported with a `Jenkins` name (or more neutrally, `Core`).

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Requesting the change as per https://github.com/jenkinsci/jenkins/pull/4385#issuecomment-574111762 . Sorry that it takes so long. I feel the PR is almost ready to go

**Comment by LinuxSuRen** [comment]
Thanks to all you. I just remove the unclear annotation for now.

**Review by fcojfernandez — APPROVED** [review]

**Review by varyvol — APPROVED** [review]

**Review by oleg-nenashev — APPROVED** [review]
Thanks!

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

**Comment by LinuxSuRen** [comment]
Thanks, everyone!

**Pull request merged**